tests_e2e/orchestrator/lib/agent_test_suite_combinator.py

```diff
@@ -0,0 +1,119 @@
+        return AgentTestSuitesCombinatorSchema
+
+    def _next(self) -> Optional[Dict[str, Any]]:
+        result: Optional[Dict[str, Any]] = None
+        if self._index < len(self._environments):
+            result = self._environments[self._index]
+            self._index += 1
+        return result
+
+    _DEFAULT_LOCATION = "westus2"
+
+    @staticmethod
+    def create_environment_list(test_suites: str) -> List[Dict[str, Any]]:
+        environment_list: List[Dict[str, Any]] = []
+        shared_environments: Dict[str, Dict[str, Any]] = {}
+
+        loader = AgentTestLoader(test_suites)
+
+        for suite_info in loader.test_suites:
+            images_info = loader.images[suite_info.images]
+            for image in images_info:
+                # If the suite specifies a location, use it. Else, if the image specifies a list of locations, use
+                # any of them. Otherwise, use the default location.
+                if suite_info.location != '':
+                    location = suite_info.location
+                elif len(image.locations) > 0:
+                    location = image.locations[0]
+                else:
+                    location = AgentTestSuitesCombinator._DEFAULT_LOCATION
+
+                # If the image specifies a list of VM sizes, use any of them. Otherwise, set the size to empty and let LISA choose it.
+                vm_size = image.vm_sizes[0] if len(image.vm_sizes) > 0 else ""
+
+                if suite_info.owns_vm:
+                    environment_list.append({
+                        "marketplace_image": image.urn,
+                        "location": location,
+                        "vm_size": vm_size,
+                        "vhd": "",
+                        "test_suites_info": [suite_info]
+                    })
+                else:
+                    key: str = f"{image.urn}:{location}"
+                    if key in shared_environments:
+                        shared_environments[key]["test_suites_info"].append(suite_info)
+                    else:
+                        shared_environments[key] = {
+                            "marketplace_image": image.urn,
+                            "location": location,
+                            "vm_size": vm_size,
+                            "vhd": "",
+                            "test_suites_info": [suite_info]
+                        }
+
+        environment_list.extend(shared_environments.values())
+
+        log: logging.Logger = logging.getLogger("lisa")
+        log.info("******** Environments *****")
+        for e in environment_list:
+            log.info(
+                "{ marketplace_image: '%s', location: '%s', vm_size: '%s', vhd: '%s', test_suites_info: '%s' }",
+                e['marketplace_image'], e['location'], e['vm_size'], e['vhd'], [s.name for s in e['test_suites_info']])
+        log.info("***************************")
+
+        return environment_list
```

Comment:
who gets this return list?

Reply:
based on this image list, lisa creates that many vms?

Reply:
The runbook uses the values returned by the combinator to feed the values used to create the VMs (see my previous comment above)